Coredump: adding an In memory backend an various improvements

tbursztyka opened the pull request:
This PR makes a few improvements in coredump sub-system, the commit titles are self-explanatory.

About the shell generalization, it gives an output like:
```
$ coredump print 
**** Zephyr Coredump ****
        Version 2
        Target: ARM Cortex-m
        Pointer size: 32
        Flag: 0
        Reason: 4
        Size of the coredump: 1884

-> Arch coredump header found
        Version 2
        Size 68
Data:
#CD:BEGIN#
#CD:0400000001000000e0282020040000004d000000112b000084c0000000000001
#CD:8828202000000000000000000000000000000000000000000000000000000000
#CD:00000000
#CD:END#

-> Thread coredump header found
        Version 1
        Size 36
Data:
#CD:BEGIN#
#CD:0000000090352020f81220207017202000000000f8122020f8122020f8122020
#CD:f8122020
#CD:END#

-> Memory coredump header found
        Version 1
        Size 200
        Starts at 0x202012f8 ends at 0x202013c0
Data:
#CD:BEGIN#
#CD:001e2020001e20200000000000800e0000000000000000000000000000000000
#CD:00000000000000000000000000000000f8122020ffffffffffffffff00000000
#CD:00000000000000000000000000000000d8292020000000005013202050132020
#CD:00010000000000000000000000000000000000002533000024f1000001000000
#CD:04000000981620207368656c6c5f756172740000000000000000000000000000
#CD:000000000000000090242020000600001000000000000000802a202000000000
#CD:0000000000000000
#CD:END#
```

tbursztyka committed:
coredump: Generalize the shell module

This might come in handy for other coredump backends too, not only flash
one. Obviously: changing the logic to use the coredump generic API instead
of the flash backend's internal functions.

Adding a better output, following how the coredump is actually created:
parsing and printing headers, and their respective content.

Signed-off-by: Tomasz Bursztyka <tobu@bang-olufsen.dk>

tbursztyka committed:
shell: kernel: Add a panic command

For testing purposes only. Testing the fatal error handler or
a coredump backend. Or just out of curiosity.

Signed-off-by: Tomasz Bursztyka <tobu@bang-olufsen.dk>

tbursztyka committed:
coredump: Have a dedicated log module for coredump

To enable/disable any coredump logging while debugging the kernel.

Signed-off-by: Tomasz Bursztyka <tobu@bang-olufsen.dk>

tbursztyka committed:
coredump: Add an in-memory backend

Only meant to be used on platforms that would not wipe the RAM on reboot.
This won't work if the platform is halted and restarted as RAM is
volatile, obviously. A relevant fatal error handler needs to be provided
to reboot.

Signed-off-by: Tomasz Bursztyka <tobu@bang-olufsen.dk>

tbursztyka committed:
include: debug: Fix indentation in coredump.h

A nano fix in the header.

Signed-off-by: Tomasz Bursztyka <tobu@bang-olufsen.dk>

mrkhldn commented on subsys/debug/coredump/coredump_shell.c:
```diff
@@ -0,0 +1,541 @@
+ * @return 0 if all has been processed, a positive value indicating the amount
+ *        that the function call has processed, a negative errno otherwise.
+ */
+static int parse_and_print_coredump(const struct shell *sh,
```
I believe that core dumps printed out in this new format will not work as expected with the existing tooling. The parser is expecting the dump in a certain format (hex print of full dump including the various headers).

https://github.com/zephyrproject-rtos/zephyr/blob/459a078d22607a0007ff3f4a2ef31c29c71a162f/scripts/coredump/coredump_serial_log_parser.py

and

https://github.com/zephyrproject-rtos/zephyr/blob/459a078d22607a0007ff3f4a2ef31c29c71a162f/scripts/coredump/coredump_gdbserver.py

tbursztyka replied:
Didn't spot these tools. There are no tests for these, it seems.

I will then add an option "pretty" to print command, so default will be the previous behavior.

tbursztyka commented on subsys/debug/coredump/coredump_backend_in_memory.c:
```diff
@@ -0,0 +1,194 @@
+
+static void coredump_in_memory_backend_start(void)
+{
+	in_memory_erase();
```
yes erasing canaries and resetting size should be enough, indeed

tbursztyka committed:
tests: debug: Add in-memory coredump backend test scenario

Providing the right config options and a new test scenario,
coredump_backends test will run under the in-memory backend.

Signed-off-by: Tomasz Bursztyka <tobu@bang-olufsen.dk>

kartben commented on subsys/debug/coredump/Kconfig:
```diff
@@ -28,6 +28,17 @@ config DEBUG_COREDUMP_BACKEND_FLASH_PARTITION
 	  Core dump is saved to a flash partition with DTS alias
 	  "coredump-partition".
 
+config DEBUG_COREDUMP_BACKEND_IN_MEMORY
```
@tbursztyka  when you have a chance, please make sure to add the two new Kconfig options to release notes for 4.2
Thanks!
